pkg/policy/snapshots.go

```diff
@@ -0,0 +1,204 @@
+package policy
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/jeremywohl/flatten"
+)
+
+func (ce *Executor) StoreSnapshot(ctx context.Context, path string, tables []string) error {
+	if len(tables) == 0 {
+		return errors.New("no tables to snapshot")
+	}
+
+	for _, table := range tables {
+		ef, err := os.OpenFile(fmt.Sprintf("%s/table_%s.csv", path, table), os.O_CREATE|os.O_WRONLY, 0777)
+		if err != nil {
+			fmt.Println("error opening file:", err)
+			return err
+		}
+		ce.log.Info("exporting", "table", table)
+
+		err = ce.conn.RawCopyTo(ctx, ef, fmt.Sprintf("COPY (select * from %s) TO STDOUT DELIMITER '|' CSV HEADER", table))
+		if err != nil {
+			return fmt.Errorf("error exporting file: %+v", err)
+		}
+		ef.Close()
+
+	}
+
+	return nil
+}
+
+func (ce *Executor) RestoreSnapshot(ctx context.Context, source string) error {
+
+	ef, err := os.OpenFile(source, os.O_RDONLY, 0777)
+	if err != nil {
+		fmt.Println("error opening file:", err)
+		return err
+	}
+	fileName := strings.Split(source, "/")
+
+	source_name := strings.TrimPrefix(strings.TrimSuffix(fileName[len(fileName)-1], ".csv"), "table_")
+	truncQuery := fmt.Sprintf("TRUNCATE %s CASCADE", source_name)
+	ce.log.Info("truncating", "table", source_name, "query", truncQuery)
+	err = ce.conn.Exec(ctx, truncQuery)
+	if err != nil {
+
+		return fmt.Errorf("error importing file: %+v", err)
+	}
+
+	query := fmt.Sprintf("copy \"%s\" from stdin DELIMITER '|' CSV HEADER;", source_name)
+	ce.log.Info("importing", "table", source_name, "query", query)
+	defer ef.Close()
+	err = ce.conn.RawCopyFrom(ctx, ef, query)
+	if err != nil {
+		return fmt.Errorf("error importing file: %+v", err)
+	}
+
+	return nil
+}
+
+func cleanQuery(query string) string {
+	var re = regexp.MustCompile(`(?s)\/\*.*?\*\/|--.*?\n`)
+	query = re.ReplaceAllString(query, "")
+	query = strings.TrimSuffix(query, ";")
+	query = strings.TrimSpace(query)
+	query = strings.TrimSuffix(query, ";")
+
+	return strings.TrimSpace(query)
+}
+func (ce *Executor) ExtractTableNames(ctx context.Context, query string) (tableNames []string, err error) {
+	ce.log.Debug("extracting Table names-raw", "raw query", query)
+	cleanedQuery := cleanQuery(query)
+	ce.log.Debug("extracting Table names-cleaned", "cleaned query", cleanedQuery)
+	explainQuery := fmt.Sprintf("EXPLAIN (FORMAT JSON) %s", cleanedQuery)
+
+	rows, err := ce.conn.Query(ctx, explainQuery)
+	if err != nil {
+		return tableNames, err
+	}
+	var s string
+	for rows.Next() {
```

Comment:
`defer rows.Close()`? not sure, check Next() godoc as some implementations close automatically